File: src/DepartmentCourse.java
import java.util.ArrayList;

public class DepartmentCourse {
    private String courseId;
    private String courseName;
    private ArrayList<int[]> courseSchedule;
    private ArrayList<Student> registeredStudents;
    private ArrayList<DepartmentCourse> prerequisites;
    private ArrayList<DepartmentCourse> prerequisiteTo;
    private ArrayList<LabCourse> labSections;
    private ArrayList<TeachingAssistant> teachingAssistants;
    private Lecturer lecturer;
    private int lectureHours;
    private int semester;
    private int credits;
    private int labHours;
    private int minCompletedCredits;
    
    public DepartmentCourse(String courseId, String courseName, int credits, int semester, int lectureHours) {
        this.courseId = courseId;
        this.courseName = courseName;
        this.credits = credits;
        this.semester = semester;
        this.lectureHours = lectureHours;
        this.registeredStudents = new ArrayList<>();
        this.prerequisites = new ArrayList<>();
        this.prerequisiteTo = new ArrayList<>();
        this.labSections = new ArrayList<>();
        this.teachingAssistants = new ArrayList<>();
        this.courseSchedule = new ArrayList<>();
    }
    public DepartmentCourse(){}

    public void addStudent(Student student) {
        registeredStudents.add(student);
    }

    public void addLectureHour(int day, int lectureHour) {
        courseSchedule.add(new int[] { day, lectureHour });
    }

    public void addPrerequisite(DepartmentCourse course) {
        prerequisites.add(course);
    }

    public void addPrerequisiteTo(DepartmentCourse course) {
        prerequisiteTo.add(course);
    }

    public void setLabHours(int labHours) {
        this.labHours = labHours;
    }

    public int labSectionsCount() {
        return labSections.size();
    }

    public void addLabSection(LabCourse labCourse) {
        labSections.add(labCourse);
    }

    public void addTeachingAssistant(TeachingAssistant teachingAssistant) {
        teachingAssistants.add(teachingAssistant);
    }

    public TeachingAssistant getTeachingAssistant(int index) {
        if (teachingAssistants.size() > index) {
            return teachingAssistants.get(index);
        } else {
            return null;
        }
    }

    public void setLecturer(Lecturer lecturer) {
        this.lecturer = lecturer;
    }

    public String getCourseId() {
        return courseId;
    }

    public String getCourseName() {
        return courseName;
    }

    public ArrayList<int[]> getCourseSchedule() {
        return courseSchedule;
    }

    public ArrayList<DepartmentCourse> getPrerequisites() {
        return prerequisites;
    }

    public ArrayList<LabCourse> getLabSections() {
        return labSections;
    }

    public int getSemester() {
        return semester;
    }

    public int getCredits() {
        return credits;
    }

    public ArrayList<Student> getRegisteredStudents() {
        return registeredStudents;
    }

    public ArrayList<DepartmentCourse> getPrerequisiteTo() {
        return prerequisiteTo;
    }

    public ArrayList<TeachingAssistant> getTeachingAssistants() {
        return teachingAssistants;
    }

    public Lecturer getLecturer() {
        return lecturer;
    }

    public int getLectureHours() {
        return lectureHours;
    }

    public int getLabHours() {
        return labHours;
    }

    public void setSemester(int semester) {
        this.semester = semester;
    }

    public void setCourseId(String courseId) {
        this.courseId = courseId;
    }

    public int getMinCompletedCredits() {
        return minCompletedCredits;
    }

    public void setMinCompletedCredits(int minCompletedCredits) {
        this.minCompletedCredits = minCompletedCredits;
    }

File: Iter3/src/department_course.py
import course
import lecturer


class DepartmentCourse(course.Course):
    __courseName: str
    __registeredStudents = []
    __prerequisites = []
    __prerequisiteTo = []
    __labSections = []
    __teachingAssistants = []

    __lecturer: type(lecturer)
    __lectureHours: int
    __semester: int
    __credits: int
    __labHours: int
    __minCompletedCredits: int

    def __init__(self, course_id, course_name, course_credits, semester, lecture_hours):
        self.courseID = course_id
        self.__courseName = course_name
        self.__credits = course_credits
        self.__semester = semester
        self.__lectureHours = lecture_hours

    def add_prerequisite(self, course):
        self.__prerequisites.append(course)

    def add_prerequisite_to(self, course):
        self.__prerequisiteTo.append(course)

    def set_lab_hours(self, lab_hours: int):
        self.__labHours = lab_hours

    def lab_sections_count(self):
        return len(self.__labSections)

    def add_lab_section(self, lab_course):
        self.__labSections.append(lab_course)

    def add_teaching_assistant(self, teaching_assistant):
        self.__teachingAssistants.append(teaching_assistant)

    def get_teaching_assistant(self, index: int):
        if len(self.__teachingAssistants) > index:
            return self.__teachingAssistants[index]
        else:
            return None

    def set_lecturer(self, lecturer):
        self.__lecturer = lecturer

    def get_course_name(self):
        return self.__courseName

    def get_prerequisites(self):
        return self.__prerequisites

    def get_lab_sections(self):
        return self.__labSections

    def get_semester(self):
        return self.__semester

    def get_credits(self):
        return self.__credits

    def get_registered_students(self):
        return self.__registeredStudents

    def get_prerequisite_to(self):
        return self.__prerequisiteTo

    def get_teaching_assistants(self):
        return self.__teachingAssistants

    def get_lecturer(self):
        return self.__lecturer

    def get_lecture_hours(self):
        return self.__lectureHours

    def get_lab_hours(self):
        return self.__labHours

    def set_semester(self, semester: int):
        self.__semester = semester

    def set_course_id(self, course_id: str):
        self.courseID = course_id

    def get_min_completed_credits(self):
        return self.__minCompletedCredits

    def set_min_completed_credits(self, min_completed: int):
        self.__minCompletedCredits = min_completed
